man/xmonad.1
.\" Automatically generated by Pandoc 3.1.3
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "XMONAD" "1" "27 October 2021" "Tiling Window Manager" ""
.hy
.SH Name
.PP
xmonad - Tiling Window Manager
.SH Description
.PP
\f[I]xmonad\f[R] is a minimalist tiling window manager for X, written in
Haskell.
Windows are managed using automatic layout algorithms, which can be
dynamically reconfigured.
At any time windows are arranged so as to maximize the use of screen
real estate.
All features of the window manager are accessible purely from the
keyboard: a mouse is entirely optional.
\f[I]xmonad\f[R] is configured in Haskell, and custom layout algorithms
may be implemented by the user in config files.
A principle of \f[I]xmonad\f[R] is predictability: the user should know
in advance precisely the window arrangement that will result from any
action.
.PP
By default, \f[I]xmonad\f[R] provides three layout algorithms: tall,
wide and fullscreen.
In tall or wide mode, windows are tiled and arranged to prevent overlap
and maximize screen use.
Sets of windows are grouped together on virtual screens, and each screen
retains its own layout, which may be reconfigured dynamically.
Multiple physical monitors are supported via Xinerama, allowing
simultaneous display of a number of screens.
.PP
By utilizing the expressivity of a modern functional language with a
rich static type system, \f[I]xmonad\f[R] provides a complete,
featureful window manager in less than 1200 lines of code, with an
emphasis on correctness and robustness.
Internal properties of the window manager are checked using a
combination of static guarantees provided by the type system, and
type-based automated testing.
A benefit of this is that the code is simple to understand, and easy to
modify.
.SH Usage
.PP
\f[I]xmonad\f[R] places each window into a \[lq]workspace\[rq].
Each workspace can have any number of windows, which you can cycle
through with mod-j and mod-k.
Windows are either displayed full screen, tiled horizontally, or tiled
vertically.
You can toggle the layout mode with mod-space, which will cycle through
the available modes.
.PP
You can switch to workspace N with mod-N.
For example, to switch to workspace 5, you would press mod-5.
Similarly, you can move the current window to another workspace with
mod-shift-N.
.PP
When running with multiple monitors (Xinerama), each screen has exactly
1 workspace visible.
mod-{w,e,r} switch the focus between screens, while shift-mod-{w,e,r}
move the current window to that screen.
When \f[I]xmonad\f[R] starts, workspace 1 is on screen 1, workspace 2 is
on screen 2, etc.
When switching workspaces to one that is already visible, the current
and visible workspaces are swapped.
.SS Flags
.PP
xmonad has several flags which you may pass to the executable.
These flags are:
.TP
\[en]recompile
Recompiles your \f[I]xmonad.hs\f[R] configuration
.TP
\[en]restart
Causes the currently running \f[I]xmonad\f[R] process to restart
.TP
\[en]replace
Replace the current window manager with xmonad
.TP
\[en]version
Display version of \f[I]xmonad\f[R]
.TP
\[en]verbose-version
Display detailed version of \f[I]xmonad\f[R]
.SS Default keyboard bindings
.TP
mod-shift-return
Launch terminal
.TP
mod-p
Launch dmenu
.TP
mod-shift-p
Launch gmrun
.TP
mod-shift-c
Close the focused window
.TP
mod-space
Rotate through the available layout algorithms
.TP
mod-shift-space
Reset the layouts on the current workspace to default
.TP
mod-n
Resize viewed windows to the correct size
.TP
mod-tab
Move focus to the next window
.TP
mod-shift-tab
Move focus to the previous window
.TP
mod-j
Move focus to the next window
.TP
mod-k
Move focus to the previous window
.TP
mod-m
Move focus to the master window
.TP
mod-return
Swap the focused window and the master window
.TP
mod-shift-j
Swap the focused window with the next window
.TP
mod-shift-k
Swap the focused window with the previous window
.TP
mod-h
Shrink the master area
.TP
mod-l
Expand the master area
.TP
mod-t
Push window back into tiling
.TP
mod-comma
Increment the number of windows in the master area
.TP
mod-period
Deincrement the number of windows in the master area
.TP
mod-shift-q
Quit xmonad
.TP
mod-q
Restart xmonad
.TP
mod-shift-slash
Run xmessage with a summary of the default keybindings (useful for
beginners)
.TP
mod-question
Run xmessage with a summary of the default keybindings (useful for
beginners)
.TP
mod-[1..9]
Switch to workspace N
.TP
mod-shift-[1..9]
Move client to workspace N
.TP
mod-{w,e,r}
Switch to physical/Xinerama screens 1, 2, or 3
.TP
mod-shift-{w,e,r}
Move client to screen 1, 2, or 3
.TP
mod-button1
Set the window to floating mode and move by dragging
.TP
mod-button2
Raise the window to the top of the stack
.TP
mod-button3
Set the window to floating mode and resize by dragging
.SH Examples
.PP
To use xmonad as your window manager add to your
\f[I]\[ti]/.xinitrc\f[R] file:
.RS
.PP
exec xmonad
.RE
.SH Customization
.PP
xmonad is customized in your \f[I]xmonad.hs\f[R], and then restarted
with mod-q.
You can choose where your configuration file lives by
.IP "1." 3
Setting \f[V]XMONAD_DATA_DIR,\f[R] \f[V]XMONAD_CONFIG_DIR\f[R], and
\f[V]XMONAD_CACHE_DIR\f[R]; \f[I]xmonad.hs\f[R] is then expected to be
in \f[V]XMONAD_CONFIG_DIR\f[R].
.IP "2." 3
Creating \f[I]xmonad.hs\f[R] in \f[I]\[ti]/.xmonad\f[R].
.IP "3." 3
Creating \f[I]xmonad.hs\f[R] in \f[V]XDG_CONFIG_HOME\f[R].
Note that, in this case, xmonad will use \f[V]XDG_DATA_HOME\f[R] and
\f[V]XDG_CACHE_HOME\f[R] for its data and cache directory respectively.
.PP
You can find many extensions to the core feature set in the xmonad-
contrib package, available through your package manager or from
xmonad.org (https://xmonad.org).
.SS Modular Configuration
.PP
As of \f[I]xmonad-0.9\f[R], any additional Haskell modules may be placed
in \f[I]\[ti]/.xmonad/lib/\f[R] are available in GHC\[cq]s searchpath.
Hierarchical modules are supported: for example, the file
\f[I]\[ti]/.xmonad/lib/XMonad/Stack/MyAdditions.hs\f[R] could contain:
.IP
.nf
\f[C]
module XMonad.Stack.MyAdditions (function1) where
  function1 = error \[dq]function1: Not implemented yet!\[dq]
\f[R]
.fi
.PP
Your xmonad.hs may then import XMonad.Stack.MyAdditions as if that
module was contained within xmonad or xmonad-contrib.
.SH Bugs
.PP
Probably.
If you find any, please report them to the
bugtracker (https://github.com/xmonad/xmonad/issues)


CHANGES.md
# Change Log / Release Notes

## _unreleased_

### Breaking Changes

### Enhancements

### Bug Fixes

## 0.18.1 (March 7, 2026)

### Breaking Changes

 * Use `cabal` for `--recompile` if there is a `.cabal` file in the config
   directory and none of `build`, `stack.yaml`, `flake.nix`, nor `default.nix`
   exist.

### Enhancements

 * Addition of `PureLayout` typeclass for `Layout` types which can work without
   any X communications. This is intended to allow XMonad's layouts to be used
   in other pluggable windowing systems (e.g., `river` and its external tiling
   support).

### Bug Fixes

### Other

PR #404 (see last change in 0.17.1) has been reverted, because the affected
compilers are (hopefully) no longer being used.

All 9.0 releases of GHC, plus 9.2.1 and 9.2.2 have the join point bug.
Note that 9.0.x is known to also have GC issues and is officially deprecated,
and the only 9.2 release that should be used is 9.2.8. Additionally, GHC HQ
doesn't support releases before 9.6.6.

## 0.18.0 (February 3, 2024)

### Breaking Changes

* Dropped support for GHC 8.4.

### Enhancements

* Exported `sendRestart` and `sendReplace` from `XMonad.Operations`.

* Exported `buildLaunch` from `XMonad.Main`.

* `Tall` does not draw windows with zero area.

* `XMonad.Operations.floatLocation` now applies size hints. This means windows
  will snap to these hints as soon as they're floated (mouse move, keybinding).
  Previously that only happened on mouse resize.

* Recompilation now detects `flake.nix` and `default.nix` (can be a
  symlink) and switches to using `nix build` as appropriate.

* Added `unGrab` to `XMonad.Operations`; this releases XMonad's passive
  keyboard grab, so other applications (like `scrot`) can do their
  thing.

### Bug Fixes

* Duplicated floats (e.g. from X.A.CopyToAll) no longer escape to inactive
  screens.

## 0.17.2 (April 2, 2023)

### Bug Fixes

  * Fixed the build with GHC 9.6.

## 0.17.1 (September 3, 2022)

### Enhancements

  * Added custom cursor shapes for resizing and moving windows.

  * Exported `cacheNumlockMask` and `mkGrabs` from `XMonad.Operations`.

  * Added `willFloat` function to `XMonad.ManageHooks` to detect whether the
    (about to be) managed window will be a floating window or not.

### Bug Fixes

  * Fixed border color of windows with alpha channel. Now all windows have the
    same opaque border color.

  * Change the main loop to try to avoid [GHC bug 21708] on systems
    running GHC 9.2 up to version 9.2.3. The issue has been fixed in
    [GHC 9.2.4] and all later releases.

[GHC bug 21708]: https://gitlab.haskell.org/ghc/ghc/-/issues/21708
[GHC 9.2.4]: https://discourse.haskell.org/t/ghc-9-2-4-released/4851

## 0.17.0 (October 27, 2021)

### Enhancements

  * Migrated `X.L.LayoutCombinators.(|||)` into `XMonad.Layout`, providing the
    ability to directly jump to a layout with the `JumpToLayout` message.

  * Recompilation now detects `stack.yaml` (can be a symlink) alongside
    `xmonad.hs` and switches to using `stack ghc`. We also updated INSTALL.md
    with instructions for cabal-install that lead to correct recompilation.

    Deprecation warnings during recompilation are no longer suppressed to make
    it easier for us to clean up the codebase. These can still be suppressed
    manually using an `OPTIONS_GHC` pragma with `-Wno-deprecations`.

  * Improve handling of XDG directories.

      1. If all three of xmonad's environment variables (`XMONAD_DATA_DIR,`
         `XMONAD_CONFIG_DIR`, and `XMONAD_CACHE_DIR`) are set, use them.
      2. If there is a build script called `build` (see [these build scripts]
         for usage examples) or configuration `xmonad.hs` in `~/.xmonad`, set
         all three directories to `~/.xmonad`.
      3. Otherwise, use the `xmonad` directory in `XDG_DATA_HOME`,
         `XDG_CONFIG_HOME`, and `XDG_CACHE_HOME` (or their respective
         fallbacks). These directories are created if necessary.

    In the cases of 1. and 3., the build script or executable is expected to be
    in the config dir.

    Additionally, the xmonad config binary and intermediate object files were
    moved to the cache directory (only relevant if using XDG or
    `XMONAD_CACHE_DIR`).

  * Added `Foldable`, `Functor`, and `Traversable` instances for `Stack`.

  * Added `Typeable layout` constraint to `LayoutClass`, making it possible to
    cast `Layout` back into a concrete type and extract current layout state
    from it.

  * Export constructor for `Choose` and `CLR` from `Module.Layout` to allow
    pattern-matching on the left and right sub-layouts of `Choose l r a`.

  * Added `withUnfocused` function to `XMonad.Operations`, allowing for `X`
    operations to be applied to unfocused windows.

[these build scripts]: https://github.com/xmonad/xmonad-testing/tree/master/build-scripts

### Bug Fixes

  * Fixed a bug when using multiple screens with different dimensions, causing
    some floating windows to be smaller/larger than the size they requested.

  * Compatibility with GHC 9.0

  * Fixed dunst notifications being obscured when moving floats.
    https://github.com/xmonad/xmonad/issues/208

### Breaking Changes

  * Made `(<&&>)` and `(<||>)` non-strict in their right operand; i.e., these
    operators now implement short-circuit evaluation so the right operand is
    evaluated only if the left operand does not suffice to determine the
    result.

  * Change `ScreenDetail` to a newtype and make `RationalRect` strict in its
    contents.

  * Added the `extensibleConf` field to `XConfig` which makes it easier for
    contrib modules to have composable configuration (custom hooks, …).

  * `util/GenerateManpage.hs` is no longer distributed in the tarball.
    Instead, the manpage source is regenerated and manpage rebuilt
    automatically in CI.

  * `DestroyWindowEvent` is now broadcasted to layouts to let them know
    window-specific resources can be discarded.

## 0.15 (September 30, 2018)


man/xmonad.hs
--
-- xmonad example config file.
--
-- A template showing all available configuration hooks,
-- and how to override the defaults in your own xmonad.hs conf file.
--
-- Normally, you'd only override those defaults you care about.
--

import XMonad
import Data.Monoid
import System.Exit

import qualified XMonad.StackSet as W
import qualified Data.Map        as M

-- The preferred terminal program, which is used in a binding below and by
-- certain contrib modules.
--
myTerminal      = "xterm"

-- Whether focus follows the mouse pointer.
myFocusFollowsMouse :: Bool
myFocusFollowsMouse = True

-- Whether clicking on a window to focus also passes the click to the window
myClickJustFocuses :: Bool
myClickJustFocuses = False

-- Width of the window border in pixels.
--
myBorderWidth   = 1

-- modMask lets you specify which modkey you want to use. The default
-- is mod1Mask ("left alt").  You may also consider using mod3Mask
-- ("right alt"), which does not conflict with emacs keybindings. The
-- "windows key" is usually mod4Mask.
--
myModMask       = mod1Mask

-- The default number of workspaces (virtual screens) and their names.
-- By default we use numeric strings, but any string may be used as a
-- workspace name. The number of workspaces is determined by the length
-- of this list.
--
-- A tagging example:
--
-- > workspaces = ["web", "irc", "code" ] ++ map show [4..9]
--
myWorkspaces    = ["1","2","3","4","5","6","7","8","9"]

-- Border colors for unfocused and focused windows, respectively.
--
myNormalBorderColor  = "#dddddd"
myFocusedBorderColor = "#ff0000"

------------------------------------------------------------------------
-- Key bindings. Add, modify or remove key bindings here.
--
myKeys conf@(XConfig {XMonad.modMask = modm}) = M.fromList $

    -- launch a terminal
    [ ((modm .|. shiftMask, xK_Return), spawn $ XMonad.terminal conf)

    -- launch dmenu
    , ((modm,               xK_p     ), spawn "dmenu_run")

    -- launch gmrun
    , ((modm .|. shiftMask, xK_p     ), spawn "gmrun")

    -- close focused window
    , ((modm .|. shiftMask, xK_c     ), kill)

     -- Rotate through the available layout algorithms
    , ((modm,               xK_space ), sendMessage NextLayout)

    --  Reset the layouts on the current workspace to default
    , ((modm .|. shiftMask, xK_space ), setLayout $ XMonad.layoutHook conf)

    -- Resize viewed windows to the correct size
    , ((modm,               xK_n     ), refresh)

    -- Move focus to the next window
    , ((modm,               xK_Tab   ), windows W.focusDown)

    -- Move focus to the next window
    , ((modm,               xK_j     ), windows W.focusDown)

    -- Move focus to the previous window
    , ((modm,               xK_k     ), windows W.focusUp  )

    -- Move focus to the master window
    , ((modm,               xK_m     ), windows W.focusMaster  )

    -- Swap the focused window and the master window
    , ((modm,               xK_Return), windows W.swapMaster)

    -- Swap the focused window with the next window
    , ((modm .|. shiftMask, xK_j     ), windows W.swapDown  )

    -- Swap the focused window with the previous window
    , ((modm .|. shiftMask, xK_k     ), windows W.swapUp    )

    -- Shrink the master area
    , ((modm,               xK_h     ), sendMessage Shrink)

    -- Expand the master area
    , ((modm,               xK_l     ), sendMessage Expand)

    -- Push window back into tiling
    , ((modm,               xK_t     ), withFocused $ windows . W.sink)

    -- Increment the number of windows in the master area
    , ((modm              , xK_comma ), sendMessage (IncMasterN 1))

    -- Deincrement the number of windows in the master area
    , ((modm              , xK_period), sendMessage (IncMasterN (-1)))

    -- Toggle the status bar gap
    -- Use this binding with avoidStruts from Hooks.ManageDocks.
    -- See also the statusBar function from Hooks.DynamicLog.
    --
    -- , ((modm              , xK_b     ), sendMessage ToggleStruts)

    -- Quit xmonad
    , ((modm .|. shiftMask, xK_q     ), io exitSuccess)

    -- Restart xmonad
    , ((modm              , xK_q     ), spawn "xmonad --recompile; xmonad --restart")

    -- Run xmessage with a summary of the default keybindings (useful for beginners)
    , ((modm .|. shiftMask, xK_slash ), xmessage help)
    ]
    ++

    --
    -- mod-[1..9], Switch to workspace N
    -- mod-shift-[1..9], Move client to workspace N
    --
    [((m .|. modm, k), windows $ f i)
        | (i, k) <- zip (XMonad.workspaces conf) [xK_1 .. xK_9]
        , (f, m) <- [(W.greedyView, 0), (W.shift, shiftMask)]]
    ++

    --
    -- mod-{w,e,r}, Switch to physical/Xinerama screens 1, 2, or 3
    -- mod-shift-{w,e,r}, Move client to screen 1, 2, or 3
    --
    [((m .|. modm, key), screenWorkspace sc >>= flip whenJust (windows . f))
        | (key, sc) <- zip [xK_w, xK_e, xK_r] [0..]
        , (f, m) <- [(W.view, 0), (W.shift, shiftMask)]]


------------------------------------------------------------------------
-- Mouse bindings: default actions bound to mouse events
--
myMouseBindings (XConfig {XMonad.modMask = modm}) = M.fromList

    -- mod-button1, Set the window to floating mode and move by dragging
    [ ((modm, button1), \w -> focus w >> mouseMoveWindow w
                                      >> windows W.shiftMaster)

    -- mod-button2, Raise the window to the top of the stack
    , ((modm, button2), \w -> focus w >> windows W.shiftMaster)

    -- mod-button3, Set the window to floating mode and resize by dragging
    , ((modm, button3), \w -> focus w >> mouseResizeWindow w
                                      >> windows W.shiftMaster)

    -- you may also bind events to the mouse scroll wheel (button4 and button5)
    ]

------------------------------------------------------------------------
-- Layouts:

-- You can specify and transform your layouts by modifying these values.
-- If you change layout bindings be sure to use 'mod-shift-space' after
-- restarting (with 'mod-q') to reset your layout state to the new
-- defaults, as xmonad preserves your old layout settings by default.
--
-- The available layouts.  Note that each layout is separated by |||,
-- which denotes layout choice.
--
myLayout = tiled ||| Mirror tiled ||| Full
  where
     -- default tiling algorithm partitions the screen into two panes
     tiled   = Tall nmaster delta ratio

     -- The default number of windows in the master pane
     nmaster = 1

     -- Default proportion of screen occupied by master pane
     ratio   = 1/2

     -- Percent of screen to increment by when resizing panes
     delta   = 3/100

------------------------------------------------------------------------
-- Window rules:

-- Execute arbitrary actions and WindowSet manipulations when managing
-- a new window. You can use this to, for example, always float a
-- particular program, or have a client always appear on a particular
-- workspace.
--
-- To find the property name associated with a program, use
-- > xprop | grep WM_CLASS
-- and click on the client you're interested in.
--
-- To match on the WM_NAME, you can use 'title' in the same way that
-- 'className' and 'resource' are used below.
--
myManageHook = composeAll
    [ className =? "MPlayer"        --> doFloat
    , className =? "Gimp"           --> doFloat
    , resource  =? "desktop_window" --> doIgnore
    , resource  =? "kdesktop"       --> doIgnore ]

------------------------------------------------------------------------
-- Event handling

-- * EwmhDesktops users should change this to ewmhDesktopsEventHook
--
-- Defines a custom handler function for X Events. The function should
-- return (All True) if the default handler is to be run afterwards. To
-- combine event hooks use mappend or mconcat from Data.Monoid.
--
myEventHook = mempty

------------------------------------------------------------------------
-- Status bars and logging

-- Perform an arbitrary action on each internal state change or X event.
-- See the 'XMonad.Hooks.DynamicLog' extension for examples.
--
myLogHook = return ()

------------------------------------------------------------------------
-- Startup hook

-- Perform an arbitrary action each time xmonad starts or is restarted
-- with mod-q.  Used by, e.g., XMonad.Layout.PerWorkspace to initialize
-- per-workspace layout choices.
--
-- By default, do nothing.
myStartupHook = return ()

------------------------------------------------------------------------
-- Now run xmonad with all the defaults we set up.

-- Run xmonad with the settings you specify. No need to modify this.
--
main = xmonad defaults

-- A structure containing your configuration settings, overriding
-- fields in the default config. Any you don't override, will
-- use the defaults defined in xmonad/XMonad/Config.hs
--
-- No need to modify this.
--
defaults = def {
      -- simple stuff
        terminal           = myTerminal,
        focusFollowsMouse  = myFocusFollowsMouse,
        clickJustFocuses   = myClickJustFocuses,
        borderWidth        = myBorderWidth,
        modMask            = myModMask,
        workspaces         = myWorkspaces,
        normalBorderColor  = myNormalBorderColor,
        focusedBorderColor = myFocusedBorderColor,

      -- key bindings
        keys               = myKeys,
        mouseBindings      = myMouseBindings,

      -- hooks, layouts
        layoutHook         = myLayout,
        manageHook         = myManageHook,
        handleEventHook    = myEventHook,
        logHook            = myLogHook,
        startupHook        = myStartupHook
    }

-- | Finally, a copy of the default bindings in simple textual tabular format.
help :: String
help = unlines ["The default modifier key is 'alt'. Default keybindings:",
    "",
    "-- launching and killing programs",
    "mod-Shift-Enter  Launch xterminal",
    "mod-p            Launch dmenu",
    "mod-Shift-p      Launch gmrun",
    "mod-Shift-c      Close/kill the focused window",
    "mod-Space        Rotate through the available layout algorithms",
    "mod-Shift-Space  Reset the layouts on the current workSpace to default",
    "mod-n            Resize/refresh viewed windows to the correct size",
    "mod-Shift-/      Show this help message with the default keybindings",
    "",
    "-- move focus up or down the window stack",
    "mod-Tab        Move focus to the next window",
    "mod-Shift-Tab  Move focus to the previous window",
    "mod-j          Move focus to the next window",
    "mod-k          Move focus to the previous window",
    "mod-m          Move focus to the master window",
    "",
    "-- modifying the window order",
    "mod-Return   Swap the focused window and the master window",
    "mod-Shift-j  Swap the focused window with the next window",
    "mod-Shift-k  Swap the focused window with the previous window",
    "",
    "-- resizing the master/slave ratio",
    "mod-h  Shrink the master area",
    "mod-l  Expand the master area",
    "",
    "-- floating layer support",
    "mod-t  Push window back into tiling; unfloat and re-tile it",
    "",
    "-- increase or decrease number of windows in the master area",
    "mod-comma  (mod-,)   Increment the number of windows in the master area",
    "mod-period (mod-.)   Deincrement the number of windows in the master area",
    "",
    "-- quit, or restart",
    "mod-Shift-q  Quit xmonad",
    "mod-q        Restart xmonad",
    "mod-[1..9]   Switch to workSpace N",
    "",
    "-- Workspaces & screens",
    "mod-Shift-[1..9]   Move client to workspace N",
    "mod-{w,e,r}        Switch to physical/Xinerama screens 1, 2, or 3",
    "mod-Shift-{w,e,r}  Move client to screen 1, 2, or 3",
    "",
    "-- Mouse bindings: default actions bound to mouse events",
    "mod-button1  Set the window to floating mode and move by dragging",
    "mod-button2  Raise the window to the top of the stack",
    "mod-button3  Set the window to floating mode and resize by dragging"]
